features/FEATURE_BLE/ble/gap/AdvertisingDataBuilder.h

```diff
@@ -516,56 +515,296 @@ class AdvertisingDataBuilder
     ble_error_t addData(
         adv_data_type_t advDataType,
         mbed::Span<const uint8_t> fieldData
-    )
-    {
-        return appendField(advDataType, fieldData);
+    ) {
+        uint8_t* field = findField(advDataType);
+
+        if (field) {
+            switch(advDataType.value()) {
+                /* These types are append to existing field */
+                case adv_data_type_t::INCOMPLETE_LIST_16BIT_SERVICE_IDS:
+                case adv_data_type_t::COMPLETE_LIST_16BIT_SERVICE_IDS:
+                case adv_data_type_t::INCOMPLETE_LIST_32BIT_SERVICE_IDS:
+                case adv_data_type_t::COMPLETE_LIST_32BIT_SERVICE_IDS:
+                case adv_data_type_t::INCOMPLETE_LIST_128BIT_SERVICE_IDS:
+                case adv_data_type_t::COMPLETE_LIST_128BIT_SERVICE_IDS:
+                case adv_data_type_t::LIST_128BIT_SOLICITATION_IDS:
+                    return appendToField(fieldData, field);
+                default:
+                    /* All other types have their field contents replaced */
+                    return replaceField(advDataType, fieldData, field);
```

Comment:
If add means add can we return an error if the field is already present ?

Reply:
addField doesn't check if the field is already there.

1. check field exists with findField()
2a). if field exists replaceField() or appendToField() for special fields
2b) if field doesn't exist addField()

In my implementation the private field functions are low level blocks that addData and removeData use.

Reply:
Usage of add and remove in addData and removeData is different but that's the public interface.